### PR TITLE
feat(slice-2b): workout-log history query endpoint with keyset paging

### PR DESCRIPTION
**Pull request by leehopper**
> Re-opened for review. This is the same change as #149, which was auto-merged on green without review and then reverted (#150). **Left open — not auto-merged.**

## What

Slice 2b **PR4 / Unit 4** — the DB-driven workout-log history query endpoint (DEC-076 § C). Adds `POST /api/v1/workouts/logs/query`: a newest-first, keyset-paginated read where sort, paging, and the page-size limit all execute as a **single SQL query** (no app-layer sort/filter/page). The response carries an opaque cursor for the next (older) page.

Reuses the existing `WorkoutLogRepository.GetByUserAsync` keyset method (`OccurredOn DESC, WorkoutLogId DESC`, `LIMIT`, user-scoped, `AsNoTracking`) shipped in PR2 — verbatim.

## Changes

- **`WorkoutLogCursorCodec`** — opaque base64 `(OccurredOn, WorkoutLogId)` token; both components are load-bearing for the keyset tiebreak, so a token that drops/corrupts either is rejected (→ 400).
- **`WorkoutLogService.QueryAsync`** — clamps the page size to `[1, 100]` (default 20), builds the next cursor when a full page is returned, and maps `WorkoutLog` → `WorkoutLogDto`.
- **`WorkoutLogsController.QueryLogs`** — `[HttpPost("query")]`, auth required, **no antiforgery** (a read, DEC-055), user-scoped via the auth claim; malformed cursor → 400 `ProblemDetails`.
- **DTOs** — `QueryWorkoutLogsRequestDto` (`limit?`, `cursor?`), `QueryWorkoutLogsResponseDto` (`logs`, `nextCursor`), `WorkoutLogDto`.
- Regenerated `swagger.json` + RTK/Zod codegen (drift gate green).

## Scope decisions (intentional deviations from the 2b spec prose)

- `WorkoutLogDto` **omits the prescription snapshot** — it feeds Slice-3 adaptation server-side, not the raw MVP-0 history view.
- The request DTO **defers the "shaped-but-unused filter block"** — the record extends additively when filter UI lands, so adding it later is not a wire break.

## Tests (25)

- `WorkoutLogCursorCodecTests` (unit) — round-trip + malformed/empty/non-base64 rejection.
- `WorkoutLogServiceTests` (unit) — page-size clamp bounds (default, floor, ceiling, passthrough).
- `WorkoutLogsControllerQueryIntegrationTests` — the eight `history-query-endpoint.feature` scenarios: newest-first, keyset paging across a boundary (no skips/dupes), exact-page-multiple termination (the last full page hands back a cursor; the trailing fetch is empty), rich-log read projection (notes/metrics/splits round-trip through `MapToDto`), empty history, per-user isolation, executed-SQL pushdown (`ORDER BY … DESC … LIMIT` via a captured EF command), and 401; plus malformed-cursor → 400.

Local: full backend suite green + all 25 new tests green; frontend build + 383 tests green; OpenAPI codegen drift gate green.

## Review follow-up (deep-review on #151)

Addressed all four deep-review suggestions (commit `cac20a2`): added the rich-log read-projection test (medium) and the exact-page-multiple termination test (low), dropped the `Slice-3` forward reference from `WorkoutLogDto`'s XML doc (low), and reconciled the contested filter-block finding by amending the Unit 4 spec to record the deferral — the request DTO grows additively when filter UI lands, so deferring avoids dead public-API surface (low).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **New Features**
  * Added workout log history query functionality enabling users to browse their past workout records.
  * Implemented cursor-based pagination for efficient browsing without retrieving all history at once.
  * Supports customizable page sizes with built-in limits to maintain performance.
  * Requires authentication to ensure users only access their own workout history.

### CODE REVIEW

File: backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs [hunk 1]
```diff
@@ -70,6 +98,30 @@ public async Task<Guid> CreateAsync(Guid userId, CreateWorkoutLogRequestDto requ
...
+            ? null
+            : JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(metrics);
+
+    private static List<WorkoutLogSplitDto>? MapSplitsToDto(IReadOnlyList<WorkoutSplit>? splits) =>
+        splits is { Count: > 0 }
+            ? splits
+                .Select(s => new WorkoutLogSplitDto(
+                    s.Index, s.DistanceMeters, s.DurationSeconds, s.PaceSecPerKm, s.AverageHeartRate))
+                .ToList()
+            : null;
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs:123]: **🟡 [MEDIUM] Read-path projection of Notes/Metrics/Splits in MapToDto is untested**

The read projection that builds the client-facing `WorkoutLogDto` — `MapToDto`, `DeserializeMetrics` (jsonb string → dictionary), and `MapSplitsToDto` — is never exercised with populated optional fields. The service unit tests mock the repository to return an empty list (so `MapToDto` is never called), and every integration seed uses the `NewLog` helper, which leaves `Notes`, `Metrics`, and `Splits` null. The create-side rich-payload tests assert on the persisted entity, not on a DTO returned through `MapToDto`. So the contract that a logged run with metrics/splits/notes is faithfully returned by the history endpoint is unverified.

**Suggested fix:** Add an integration test that seeds a log with `Notes`, a `Metrics` jsonb (e.g. `{"hrAvg":142}`), and two splits, queries it, and asserts the returned `WorkoutLogDto` echoes `Notes`, deserializes `Metrics` to the expected keys/values, and maps `Splits` with correct `Index`/`DistanceMeters`/`PaceSecPerKm`/`AverageHeartRate`.

_Confidence: 67% · deep-review (improvement suggestion)_

File: backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs [hunk 2]
```diff
@@ -59,6 +65,28 @@ public async Task<Guid> CreateAsync(Guid userId, CreateWorkoutLogRequestDto requ
...
+
+        // A full page means there may be more, so hand back the tail as the next
+        // cursor; a short or empty page is the last one. Detecting "more" by row
+        // count — rather than fetching limit+1 and trimming — keeps every returned
+        // row exactly as the DB ordered and trimmed it (no app-layer slicing), at
+        // the cost of one extra empty fetch when the total is an exact multiple of
+        // the page size.
+        var nextCursor = logs.Count == limit
+            ? WorkoutLogCursorCodec.Encode(new WorkoutLogCursor(logs[^1].OccurredOn, logs[^1].WorkoutLogId))
+            : null;
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs:84]: **💡 [LOW] Untested exact-page-multiple cursor termination**

`QueryAsync` detects more pages by row count: when `logs.Count == limit` it emits a non-null `nextCursor`, otherwise null. When the total log count is an exact multiple of the page size, the last full page still returns a non-null cursor and a subsequent fetch returns zero rows with a null cursor (the code comment documents this trade-off). The only multi-page paging test seeds 5 logs at limit 2 (pages 2/2/1 — a short final page), so the exact-multiple branch is never exercised.

**Suggested fix:** Add a test seeding 4 logs at limit 2: assert page1 (2 logs, non-null cursor), page2 (2 logs, **non-null** cursor), then a page3 fetch using page2's cursor returns 0 logs and a null cursor — with no log duplicated or skipped across the empty terminal page.

_Confidence: 90% · deep-review (improvement suggestion)_

File: backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogDto.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace RunCoach.Api.Modules.Training.Workouts;
+
+/// <summary>
+/// Read projection of a logged workout for the history surface (slice-2b Unit 4).
+/// Mirrors the actual-run fields the runner recorded: required core fields, the
+/// open optional-metrics bag (rehydrated from the entity's <c>jsonb</c> string),
+/// the freeform note, and the typed display-only splits. The server-authoritative
+/// prescription snapshot is intentionally not surfaced here at MVP-0 — it feeds
+/// Slice-3 deterministic adaptation server-side, not the raw history view.
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogDto.cs]: **💡 [LOW] Comment contains a forward reference to future Slice-3 work**

The XML-doc justifies the omitted prescription snapshot with a forward reference to future-slice behavior: "it feeds Slice-3 deterministic adaptation server-side, not the raw history view." REVIEW.md's code-comments rule bans planning-phase forward references (examples: "Slice 1 only", "later slices will"). This is borderline — the comment is partly load-bearing (it justifies the intentional MVP-0 omission, which the rule permits) — but the `Slice-3` phrasing matches the prohibited shape.

**Suggested fix:** Keep the load-bearing why, drop the forward-looking slice reference — e.g. "the snapshot feeds server-side deterministic adaptation, not the raw history view" — or cite the governing DEC instead of the slice number.

_Confidence: 57% · deep-review (improvement suggestion)_

File: backend/src/RunCoach.Api/Modules/Training/Workouts/QueryWorkoutLogsRequestDto.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+/// Keyset paging only at MVP-0: an optional opaque <paramref name="Cursor"/> (a
+/// prior page's <c>nextCursor</c>; null/absent starts at the most recent log) and
+/// an optional <paramref name="Limit"/> page size (clamped server-side). Ordering
+/// is fixed newest-first (<c>OccurredOn</c> desc, id tiebreak). A filter block is
+/// intentionally deferred — the record grows additively when filter UI arrives, so
+/// adding it later is not a wire break.
+/// </summary>
+/// <param name="Limit">Requested page size; null applies the server default. Clamped to the server's [1, max] bound.</param>
+/// <param name="Cursor">Opaque keyset cursor from a prior page's <c>nextCursor</c>; null/empty starts at the newest log.</param>
+public sealed record QueryWorkoutLogsRequestDto(int? Limit, string? Cursor);
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Training/Workouts/QueryWorkoutLogsRequestDto.cs:14]: **💡 [LOW] Request DTO omits the filter block the spec said to shape now (contested)**

Spec Unit 4 functional requirement (line 79) states the query DTO "shall accept … an optional filter block, shaped-but-unused at MVP-0" — i.e. the filter shape should be present now even though no filter logic runs yet. The implemented `QueryWorkoutLogsRequestDto` carries only `Limit` and `Cursor`; the comment explicitly defers the filter block.

This finding is **contested**: the validator scored it low (the spec's out-of-scope note "the query DTO is shaped to grow" and the documented additive-growth approach support deferral, and the binding BDD acceptance criteria have no filter scenario), while the blind challenger read the normative "shall … shaped-but-unused at MVP-0" wording as mandating the shape now. Flagging so spec and code can be reconciled.

**Suggested fix:** Either add the (empty) optional filter block to the request DTO to satisfy line 79, or amend the Unit 4 spec to record that the filter shape is deferred (consistent with the line-135 growth note) so spec and code agree.

_Confidence: 15% (contested) · deep-review (improvement suggestion)_

**Review by leehopper — COMMENTED** [review]
## Deep Review — PR #151

Reviewed with the deep-review multi-agent pipeline (Frontier mode: 7 Opus discovery agents → deterministic verification → Sonnet validation → Opus blind challenge).

**This is a clean PR — 0 blocking issues.** Correctness, security, cross-file impact, and type design all came back clean after verification and blind challenge. The blind-challenge round also removed one false positive (a keyset tie-break gap that `WorkoutLogRepositoryIntegrationTests.GetByUser_KeysetCursor_SameDate_PagesViaIdTiebreakWithoutSkipsOrDuplicates` already covers end-to-end).

Posting **4 improvement suggestions** as inline comments — 1 medium and 3 low, all non-blocking:

- 🟡 Read-path projection (`Notes`/`Metrics`/`Splits`) of `MapToDto` is untested
- 💡 Exact-page-multiple cursor termination is untested
- 💡 Comment forward-reference to Slice-3 (REVIEW.md code-comments rule)
- 💡 Request DTO omits the spec's "shaped-but-unused" filter block (contested)

_Generated by deep-review · Reviewed up to: ffc7baf1c89c740a3315340d827ae8e955037744_

File: backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs (hunk 1 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs:123]: Addressed in cac20a2. Added `Query_RichLog_EchoesNotesMetricsAndSplitsThroughTheReadProjection`: it seeds a log with `Notes`, a `{"hrAvg":142,"rpe":7}` metrics jsonb, and two splits, queries the endpoint, and asserts the returned `WorkoutLogDto` echoes `Notes`, deserializes `Metrics` to the expected keys/values, and maps both `Splits` in order (`Index`/`DistanceMeters`/`DurationSeconds`/`PaceSecPerKm`/`AverageHeartRate`). This exercises `MapToDto` / `DeserializeMetrics` / `MapSplitsToDto` end-to-end through the HTTP read path. Green locally.

File: backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs (hunk 2 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogService.cs:84]: Addressed in cac20a2. Added `Query_ExactPageMultiple_TerminatesWithEmptyTrailingPage`: seeds 4 logs at limit 2, then asserts page1 (2 logs, non-null cursor), page2 (2 logs, **non-null** cursor), and a page3 fetch on page2's cursor returning 0 logs with a null cursor — with no log skipped or duplicated across the empty terminal page. Exercises the exact-page-multiple branch.

File: backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogDto.cs (hunk 3 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Training/Workouts/WorkoutLogDto.cs]: Addressed in cac20a2. Dropped the `Slice-3` forward reference per the REVIEW.md code-comments rule and kept the load-bearing why — the XML doc now reads "it feeds server-side deterministic adaptation, not the raw history view."

File: backend/src/RunCoach.Api/Modules/Training/Workouts/QueryWorkoutLogsRequestDto.cs (hunk 4 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Training/Workouts/QueryWorkoutLogsRequestDto.cs:14]: Resolved in favor of deferral (the "amend the spec" option). The request DTO is a C# record that gains an optional filter property *additively* when filter UI lands — that is not a wire break — so pre-shaping the block now would add dead public-API surface (it would appear in `swagger.json` and the generated Zod/RTK types) that nothing consumes at MVP-0.

I amended the Unit 4 spec (commit cac20a2) to remove the contradictory "shaped-but-unused filter block" wording so it now agrees with the existing "query DTO is shaped to grow; no filter UI at MVP-0" non-goal and with the code. No code change.
